File: proxy/nodejs/src/routes/execute.js
const express = require('express');
const router = express.Router();
const { getAdapter } = require('../db/connector');
const { loadConfig } = require('./config');
const { writeAuditLog } = require('../utils/auditLogger');

function splitSql(sql) {
  const results = [];
  let current = '';
  let inString = false;
  let stringChar = '';
  for (let i = 0; i < sql.length; i++) {
    const ch = sql[i];
    if (inString) {
      current += ch;
      if (ch === stringChar && sql[i - 1] !== '\\') inString = false;
    } else if (ch === "'" || ch === '"') {
      inString = true;
      stringChar = ch;
      current += ch;
    } else if (ch === ';') {
      const trimmed = current.trim();
      if (trimmed) results.push(trimmed);
      current = '';
    } else {
      current += ch;
    }
  }
  const trimmed = current.trim();
  if (trimmed) results.push(trimmed);
  return results;
}

function parseSqls(body) {
  if (Array.isArray(body.sqls)) return body.sqls.filter(s => s && s.trim());
  if (typeof body.sql === 'string') return splitSql(body.sql);
  return [];
}

// POST /execute - 단일 SQL 실행
router.post('/', async (req, res) => {
  const config = loadConfig();
  if (!config) return res.status(400).json({ error: '접속정보가 설정되지 않았습니다.' });

  const sql = req.body.sql;
  if (!sql || !sql.trim()) return res.status(400).json({ error: 'SQL이 비어있습니다.' });

  try {
    const adapter = getAdapter(config.dbType);
    const start = Date.now();
    const result = await adapter.execute(config, sql.trim());
    const duration = Date.now() - start;
    writeAuditLog('EXECUTE', sql.trim(), { durationMs: duration, rowCount: result.rowCount });
    res.json({ ok: true, ...result, duration });
  } catch (err) {
    writeAuditLog('EXECUTE', sql.trim(), { error: err.message });
    res.status(400).json({ ok: false, error: err.message });
  }
});

// POST /execute/stream - 다중 SQL SSE 스트리밍
router.post('/stream', async (req, res) => {
  const config = loadConfig();
  if (!config) {
    res.status(400).json({ error: '접속정보가 설정되지 않았습니다.' });
    return;
  }

  const sqls = parseSqls(req.body);
  if (sqls.length === 0) {
    res.status(400).json({ error: '실행할 SQL이 없습니다.' });
    return;
  }

  const stopOnError = req.body.stopOnError === true;

  // SSE 헤더 설정
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();

  function send(event, data) {
    res.write(`event: ${event}\ndata: ${JSON.stringify(data)}\n\n`);
  }

  const adapter = getAdapter(config.dbType);
  const total = sqls.length;
  let success = 0;
  let failed = 0;
  const startAll = Date.now();

  for (let i = 0; i < sqls.length; i++) {
    const sql = sqls[i];
    const step = i + 1;

    send('progress', { step, total, sql, status: 'running' });

    try {
      const start = Date.now();
      const result = await adapter.execute(config, sql);
      success++;
      const duration = Date.now() - start;
      writeAuditLog('STREAM', sql, { durationMs: duration, rowCount: result.rowCount });
      send('progress', {
        step, total, sql, status: 'ok',
        rowCount: result.rowCount,
        duration
      });
    } catch (err) {
      failed++;
      writeAuditLog('STREAM', sql, { error: err.message });
      send('error', { step, total, sql, error: err.message });
      if (stopOnError) break;
    }
  }

  send('done', { success, failed, total, duration: Date.now() - startAll });
  res.end();
});

module.exports = router;

File: proxy/python/routers/execute.py
import json
import time
from typing import List, Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from db.connector import get_adapter
from routers.config import load_config
from utils.audit_logger import write_audit_log

router = APIRouter()


def _split_sql(sql: str) -> List[str]:
    results = []
    current = ""
    in_string = False
    string_char = ""
    for i, ch in enumerate(sql):
        if in_string:
            current += ch
            if ch == string_char and (i == 0 or sql[i - 1] != "\\"):
                in_string = False
        elif ch in ("'", '"'):
            in_string = True
            string_char = ch
            current += ch
        elif ch == ";":
            trimmed = current.strip()
            if trimmed:
                results.append(trimmed)
            current = ""
        else:
            current += ch
    trimmed = current.strip()
    if trimmed:
        results.append(trimmed)
    return results


def _parse_sqls(body_sqls, body_sql) -> List[str]:
    if body_sqls is not None:
        return [s for s in body_sqls if s and s.strip()]
    if body_sql is not None:
        return _split_sql(body_sql)
    return []


# ── POST /execute ─────────────────────────────────────────────────────────────

class ExecuteBody(BaseModel):
    sql: str


@router.post("")
@router.post("/")
async def execute_sql(body: ExecuteBody):
    config = load_config()
    if not config:
        raise HTTPException(status_code=400, detail="접속정보가 설정되지 않았습니다.")
    if not body.sql or not body.sql.strip():
        raise HTTPException(status_code=400, detail="SQL이 비어있습니다.")
    try:
        adapter = get_adapter(config["dbType"])
        start = time.time()
        result = await adapter.execute(config, body.sql.strip())
        duration = int((time.time() - start) * 1000)
        write_audit_log("EXECUTE", body.sql.strip(), {"durationMs": duration, "rowCount": result.get("rowCount", 0)})
        return {"ok": True, **result, "duration": duration}
    except Exception as e:
        write_audit_log("EXECUTE", body.sql.strip(), {"error": str(e)})
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail={"ok": False, "error": str(e)})


# ── POST /execute/stream ──────────────────────────────────────────────────────

class StreamBody(BaseModel):
    sqls: Optional[List[str]] = None
    sql: Optional[str] = None
    stopOnError: Optional[bool] = False


@router.post("/stream")
async def execute_stream(body: StreamBody):
    config = load_config()
    if not config:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="접속정보가 설정되지 않았습니다.")
    sqls = _parse_sqls(body.sqls, body.sql)
    if not sqls:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail="실행할 SQL이 없습니다.")
    stop_on_error = body.stopOnError or False

    async def event_generator():
        adapter = get_adapter(config["dbType"])
        total = len(sqls)
        success = 0
        failed = 0
        start_all = time.time()

        for i, sql in enumerate(sqls):
            step = i + 1
            yield f"event: progress\ndata: {json.dumps({'step': step, 'total': total, 'sql': sql, 'status': 'running'}, ensure_ascii=False)}\n\n"
            try:
                start = time.time()
                result = await adapter.execute(config, sql)
                success += 1
                duration = int((time.time() - start) * 1000)
                write_audit_log("STREAM", sql, {"durationMs": duration, "rowCount": result.get("rowCount", 0)})
                yield f"event: progress\ndata: {json.dumps({'step': step, 'total': total, 'sql': sql, 'status': 'ok', 'rowCount': result.get('rowCount', 0), 'duration': duration}, ensure_ascii=False)}\n\n"
            except Exception as e:
                failed += 1
                write_audit_log("STREAM", sql, {"error": str(e)})
                yield f"event: error\ndata: {json.dumps({'step': step, 'total': total, 'sql': sql, 'error': str(e)}, ensure_ascii=False)}\n\n"
                if stop_on_error:
                    break

        total_duration = int((time.time() - start_all) * 1000)
        yield f"event: done\ndata: {json.dumps({'success': success, 'failed': failed, 'total': total, 'duration': total_duration}, ensure_ascii=False)}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "Connection": "keep-alive"},
    )
